src/consensus/produce_block.rs

```diff
@@ -113,63 +114,112 @@ where
-            parent_ready,
         );
 
         let mut sleep_duration = DELTA_BLOCK;
-        for slice_index in 0.. {
-            let slice_index = match NonZeroUsize::new(slice_index) {
-                None => Either::Left((parent_slot, parent_hash, 0)),
-                Some(ind) => Either::Right(ind),
+        let mut parent_changed = None;
+        for ind in 0.. {
+            let slice_index = match (ind, parent_changed.take()) {
+                (0, None) => Either::Left((parent_slot, parent_hash, 0)),
+                (0, _) => unreachable!(),
+                (ind, Some((new_slot, new_hash))) => Either::Left((new_slot, new_hash, ind)),
+                (ind, None) => Either::Right(NonZeroUsize::new(ind).unwrap()),
             };
-            let (slice, cont_prod) =
-                produce_slice(&self.txs_receiver, slot, slice_index, sleep_duration).await;
-            // shred and disseminate slice
-            let shreds = RegularShredder::shred(slice, &self.secret_key).unwrap();
-            for s in shreds {
-                self.disseminator.send(&s).await?;
-                // PERF: move expensive add_shred() call out of block production
-                let block = self
-                    .blockstore
-                    .write()
-                    .await
-                    .add_shred_from_disseminator(s)
-                    .await;
-                if let Ok(Some((slot, block_info))) = block {
-                    self.pool.write().await.add_block(slot, block_info).await;
-                }
-            }
 
-            // switch parent if necessary (for optimistic handover)
-            if !parent_ready {
-                let pool = self.pool.read().await;
-                if let Some(p) = pool.parents_ready(slot).first()
-                    && *p != parent
-                {
-                    warn!(
-                        "switching block production parent from {} in slot {} to {} in slot {}",
-                        &hex::encode(parent.1)[..8],
-                        parent.0,
-                        &hex::encode(p.1)[..8],
-                        p.0,
-                    );
-                    unimplemented!("have to switch parents");
+            let future = produce_slice(&self.txs_receiver, slot, slice_index, sleep_duration);
+            let (mut slice, cont_prod) = if receiver.is_terminated() {
+                future.await
+            } else {
+                pin!(future);
+                tokio::select! {
+                    res = &mut future => res,
+                    res = &mut receiver => {
+                        let (new_slot, new_hash) = res.unwrap();
+                        if new_slot != parent_slot && new_hash != parent_hash {
+                            parent_changed = Some((new_slot, new_hash));
+                        }
+                        future.await
+                  }
                 }
+            };
+            if let Continue::Stop = &cont_prod
```

Comment:
- Also delta block should start only after we see parent ready
- introduce max slices per block to start accounting for CUs.
- in current version: need to ensure that parent ready has fired here.